libraries/tickv/README.md

```diff
@@ -0,0 +1,110 @@
+# TicKV
+
+TicKV (Tiny Circular Key Value) is a small file system allowing
+key value pairs to be stored in Flash Memory.
+
+TicKV was written to allow the Tock OS kernel to persistently store app data
+on flash. It was written to be generic though, so other Rust applications can
+use it if they want.
+
+TicKV is based on similar concepts as
+[Yaffs1](https://yaffs.net/documents/how-yaffs-works]).
+
+## Goals of TicKV
+
+TicKV is designed with these main goals (in order)
+
+ * Fully implemented in no_std Rust
+ * Power loss resilient
+ * Maintain data integrity and detect media errors
+ * Wear leveling
+ * Low memory usage
+ * Low storage overhead
+ * No external crates in use (not including unit tests)
+
+TicKV is also designed with some assumptions
+
+ * Most operations will be retrieving keys
+ * Some operations will be storing keys
+ * Keys will rarely be deleted
+ * Key values will rarely need to be modified
+
+## ACID characteristics
+
+TicKV provides some ACID properties. For the purpose of ACID a transaction is
+a key operation, that is finding, adding, invalidating or fully removing
+(garbage collection) a key.
+
+Atomicity: TicKV does not provide atomicity. If a operation fails mid write
+then there will be an invalid key/value stored to flash.
+
+Consistency: Consistency is maintained similar to atomicity. All operations
```

Comment:
What does "similar to atomicity" mean? The atomicity comment says a database can have an invalid key/value. So it is possible for an operation to take the store into an invalid state?

It is useful to state when it *is* atomic or consistent. E.g., you can say that "Under the constraint that underlying flash operations succeed, TicKV provides atomicity. If the underlying flash has an error, the system crashes mid-operation, or power is lost mid-operation, an operation may be partially committed."

Reply:
What happens if a write fails? Is the rest of the page considered full?

Reply:
I have updated this in the doc to answer all your questions. Let me know what you think. I have pasted it below as well:

> ## System Guarantees
> 
> A successful write or erase is durably committed.
> 
> The design does not support concurrency, such that it imposes a total order
> on all read, write and delete operations. Successful individual operations
> are therefore atomic. Applications that require higher-level atomicity
> (e.g., read/modify/delete/write, multiple writes) need to build this on top
> of these operations.
> 
> TicKV is not robust to low-level flash failures, power loss, or system
> crashes. However, a failure only affects a single key: a failure to write
> or delete key K will corrupt at most only K. A value to read will have no
> effect.
> 
> Under the constraint that underlying flash operations succeed, TicKV provides
> atomicity. If the underlying flash has an error, the system crashes
> mid-operation, or power is lost mid-operation, an operation may be
> partially committed.
> 
> If a write fails with only a partial write to flash the entire length of
> the key will be used in the storage.
> 
> TicKV ensures durability and once a transaction has completed
> and been committed to flash it will remain there. TicKV also takes measures
> to apply wear levelling to the flash storage.